K8s: fixed issues related to links / ownership links

hunchback opened the pull request:
Make improvements and fixes to the following links:
- pod --ownership--> node
- container --ownership--> pod
- container --association--> docker
- networkPolicy --association--> pod

hunchback commented:
@lebauce / @safchain - kindly review ...

lebauce commented on topology/probes/k8s/client.go:
```diff
@@ -80,6 +81,41 @@ type kubeCache struct {
+	return f.String()
+}
+
+func (c *kubeCache) listByNamespace(namespace string) (objList []interface{}) {
```
It seems this function is used only once, and the type is perfectly known so there is no reason to use a slow reflection to get a simple field.

hunchback replied:
Is is going to be used by other throughout the code base once we create graph.Node to represent namespaces as in:

```
$ kubectl get namespaces
NAME          STATUS    AGE
default       Active    3d
kube-public   Active    3d
kube-system   Active    3d
```

lebauce replied:
I suggest we do this modification when we'll really need it.

hunchback replied:
ok, will be addressed in update to: `k8s: networkPolicy chain namespace-filter then pod-selection-filter`

lebauce commented on topology/probes/k8s/graph.go:
```diff
@@ -24,22 +24,58 @@ package k8s
 	"github.com/skydive-project/skydive/topology/graph"
 )
 
+func setManager(m *graph.Metadata) {
```
Maps are passed as reference so there's not need to pass it as a pointer.

Besides, it's a bit overkill to create a function just to set an attribute. Maybe defining a constant for the manager is enough (KubernetesManager ?) and set it inline where setManager is called.

hunchback replied:
agreed x 2 ... will fix accordingly

lebauce commented on topology/probes/k8s/container.go:
```diff
@@ -86,16 +86,35 @@ func containerUID(pod *v1.Pod, containerName string) graph.Identifier {
 	return graph.GenIDNameBased(string(pod.GetUID()), containerName)
 }
 
+func dumpContainer2(podNamespace, podName, containerName string) string {
```
`func dumpContainer(pod *v1.Pod, containerName string) string` may be enough

hunchback replied:
I wanted to have a dump for containers when we don't have readily available kubernetes objects (such as v1.Pod, etc.)

lebauce replied:
For the only call to dumpContainer2 (line 175 of container.go), we have v1.Pod object

hunchback replied:
ok, will be addressed in: `k8s: improve logging of CRUD operations`

hunchback commented on topology/probes/k8s/networkpolicy.go:
```diff
@@ -46,134 +48,111 @@ func (n *networkPolicyProbe) newMetadata(np *networking_v1.NetworkPolicy) graph.
-		} else {
-			n.graph.Unlink(policyNode, podNode)
-		}
+		syncLink(n.graph, policyNode, podNode, n.isPodSelected(policy, pod))
```
Have fixed syncLink so now it either added a link or deletes a link (thus reverting to the original behavior of the code).

Please see commit named: `k8s: add utility functions for handling graph.Edge`

hunchback commented on topology/topology.go:
```diff
@@ -119,21 +119,36 @@ func IsOwnershipLinked(g *graph.Graph, node *graph.Node) bool {
-	// a child node can only have one parent of type ownership, so delete the previous link
-	for _, e := range g.GetNodeEdges(child, graph.Metadata{"RelationType": OwnershipLink}) {
+// GetLink get Link between the parent and the child node or nil
+func GetLink(g *graph.Graph, parent *graph.Node, child *graph.Node, metadata graph.Metadata) *graph.Edge {
```
ok, will address in update to: `k8s: add utility functions for handling graph.Edge`

hunchback commented on topology/probes/k8s/networkpolicy.go:
```diff
@@ -46,134 +48,111 @@ func (n *networkPolicyProbe) newMetadata(np *networking_v1.NetworkPolicy) graph.
+	n.handleNetworkPolicyAdd(obj)
 }
 
 func (n *networkPolicyProbe) OnUpdate(oldObj, newObj interface{}) {
```
ok have returned the prior behavior, it was replaced not sue to a specific but but so as to simplify the logic, but as said now reverted

hunchback commented:
@lebauce - I have uploaded a new version fixing all identified issues - please review

hunchback commented:
@lebauce, @safchain - I have addressed setting hostID to "" (for Node/Edge elements):

```
topology: add ability to set host in g.Link() AddOwnershipLink()
k8s: add/fix newNode(), link(), addOwnershipLink() with empty hostID
k8s: call new newNode() which sets emtpry hostID
flow: fixed bad gofmt
```

safchain commented:
ok to test

lebauce commented:
add to whitelist

safchain commented:
rerun functional-tests-backend-elasticsearch
